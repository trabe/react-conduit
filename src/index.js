export { default as ConduitProvider } from "./conduit-provider";
export { default as Inlet } from "./inlet";
export { default as Outlet } from "./outlet";
export { createRegistry } from "./registry";
