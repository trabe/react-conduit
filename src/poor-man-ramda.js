/* global Reflect */
export const values = obj => Object.values(obj);

export const assoc = (k, v, obj) => ({ ...obj, [k]: v });

export const dissoc = (k, obj) => {
  const ret = { ...obj };
  Reflect.deleteProperty(ret, k);
  return ret;
};
