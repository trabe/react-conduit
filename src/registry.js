import createFragment from "react-addons-create-fragment";
import uuidV4 from "uuid";

export const createRegistry = () => ({
  // outletId -> { id: outletId, inlets: [inletIds], watchers: { uuid: fn }}
  outlets: {},
  // inletId -> children
  children: {},
});

const initializeOutlet = (registry, { id, inlets = new Set(), watchers = {} }) => {
  registry.outlets[id] = { id, inlets, watchers };
};

const ensureOutletInitialized = fn =>
  (registry, outletId, ...args) => {
    if (!registry.outlets[outletId]) {
      initializeOutlet(registry, { id: outletId });
    }
    return fn(registry, outletId, ...args);
  };

export const watchOutlet = ensureOutletInitialized((registry, outletId, fn) => {
  const watcherId = uuidV4();
  registry.outlets[outletId].watchers[watcherId] = fn;

  return () => {
    delete registry.outlets[outletId].watchers[watcherId];
  };
});

export const mergeInletChildren = ({ outlets, children }, outletId) =>
  createFragment(
    Array.from((outlets[outletId] && outlets[outletId].inlets) || [])
      .map(inletId => [inletId, children[inletId]])
      .reduce((acc, [id, children]) => ({ ...acc, [id]: children }), {}),
  );

export const addInlet = ensureOutletInitialized((registry, outletId, inletId) => {
  registry.outlets[outletId].inlets.add(inletId);
  notifyWatchers(registry, outletId);
});

export const removeInlet = ensureOutletInitialized((registry, outletId, inletId) => {
  registry.outlets[outletId].inlets.delete(inletId);
  notifyWatchers(registry, outletId);
});

export const updateChildren = (registry, inletId, children) => {
  registry.children[inletId] = children;

  Object.values(registry.outlets)
    .filter(outlet => outlet.inlets.has(inletId))
    .forEach(outlet => notifyWatchers(registry, outlet.id));
};

export const notifyWatchers = ensureOutletInitialized(({ outlets }, outletId) => {
  Object.values(outlets[outletId].watchers).forEach(watcher => watcher());
});
