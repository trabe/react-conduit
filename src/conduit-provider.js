import React, { Component, Children, PropTypes } from "react";
import { createRegistry } from "./registry";

class ConduitProvider extends Component {
  constructor(props) {
    super(props);
    const { registry } = this.props;

    this.registry = registry ? registry : createRegistry();
  }

  getChildContext() {
    return { registry: this.registry };
  }

  render() {
    return Children.only(this.props.children);
  }
}

ConduitProvider.childContextTypes = {
  registry: PropTypes.object.isRequired,
};

export default ConduitProvider;
