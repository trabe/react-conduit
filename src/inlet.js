import React, { Component, Children, PropTypes } from "react";
import uuidV4 from "uuid/v4";
import { updateChildren, removeInlet, addInlet } from "./registry";

export class Inlet extends Component {
  constructor(props) {
    super(props);
    this.id = uuidV4();
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.label !== nextProps.label) {
      removeInlet(this.context.registry, this.props.label, this.id);
      addInlet(this.context.registry, nextProps.label, this.id);
    }
    updateChildren(this.context.registry, this.id, nextProps.children);
  }

  componentWillMount() {
    addInlet(this.context.registry, this.props.label, this.id);
    updateChildren(this.context.registry, this.id, this.props.children);
  }

  componentWillUnmount() {
    removeInlet(this.context.registry, this.props.label, this.id);
  }

  render() {
    return null;
  }
}

Inlet.contextTypes = {
  registry: PropTypes.object.isRequired,
};

Inlet.PropTypes = {
  label: PropTypes.string.isRequired,
};

export default Inlet;
