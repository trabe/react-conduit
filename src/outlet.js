import React, { Component, PropTypes } from "react";
import { watchOutlet, mergeInletChildren } from "./registry";

class Outlet extends Component {
  constructor(props, context) {
    super(props, context);

    this.state = { children: this.getChildren() };

    this.getChildren = this.getChildren.bind(this);
    this.updateChildren = this.updateChildren.bind(this);
  }

  componentWillMount() {
    watchOutlet(this.context.registry, this.props.label, this.updateChildren);
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.label !== nextProps.label) {
      this.updateChildren();
    }
  }

  getChildren() {
    return mergeInletChildren(this.context.registry, this.props.label);
  }

  updateChildren() {
    this.setState({ children: this.getChildren() });
  }

  render() {
    return (
      <div>
        {this.state.children}
      </div>
    );
  }
}

Outlet.contextTypes = {
  registry: PropTypes.object.isRequired,
};

Outlet.PropTypes = {
  label: PropTypes.string.isRequired,
};

export default Outlet;
