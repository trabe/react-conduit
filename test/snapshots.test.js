import React from "react";
import { render } from "enzyme";
import { ConduitProvider, Inlet, Outlet } from "../src";

const renderWithProvider = element =>
  render(
    <ConduitProvider>
      {element}
    </ConduitProvider>,
  );

describe("conduit snapshots", () => {
  describe("Inlet", () => {
    it("Inlet should not render its children", () => {
      const wrapper = renderWithProvider(
        <Inlet label="outlet1">
          <div>this shouldn't be seen</div>
        </Inlet>,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Outlet", () => {
    it("Outlet should not render its children", () => {
      const wrapper = renderWithProvider(
        <Outlet label="outlet1">
          <div>this shouldn't be seen</div>
        </Outlet>,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Inlet and Outlet", () => {
    it("should output Inlet's children to the corresponding outlet", () => {
      const wrapper = renderWithProvider(
        <div>
          <Inlet label="1">this should be rendered in the outlet</Inlet>

          <Outlet label="1" />
        </div>,
      );
      expect(wrapper).toMatchSnapshot();
    });

    it("should merge the associated Inlet's children to the corresponding outlet", () => {
      const wrapper = renderWithProvider(
        <div>
          <Inlet label="1"><div>first inlet</div></Inlet>
          <Inlet label="1"><div>second inlet</div></Inlet>
          <Inlet label="1"><div>third inlet</div></Inlet>

          <Outlet label="1" />
        </div>,
      );
      expect(wrapper).toMatchSnapshot();
    });

    it("Inlet children are sent to their corresponding outlets", () => {
      const wrapper = renderWithProvider(
        <div>
          <Inlet label="1"><div>first inlet</div></Inlet>
          <Inlet label="1"><div>second inlet</div></Inlet>
          <Inlet label="2"><div>third inlet</div></Inlet>

          <div id="outlet1">
            <Outlet label="1" />
          </div>
          <div id="outlet2">
            <Outlet label="2" />
          </div>
        </div>,
      );
      expect(wrapper).toMatchSnapshot();
    });

    it("two outlets with the same label will duplicate the children rendering", () => {
      const wrapper = renderWithProvider(
        <div>
          <Inlet label="1"><div>inlet</div></Inlet>

          <Outlet label="1" />
          <Outlet label="1" />
        </div>,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
