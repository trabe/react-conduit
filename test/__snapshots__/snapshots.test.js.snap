// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conduit snapshots Inlet Inlet should not render its children 1`] = `null`;

exports[`conduit snapshots Inlet and Outlet Inlet children are sent to their corresponding outlets 1`] = `
<div>
  <div
    id="outlet1"
  >
    <div>
      <div>
        first inlet
      </div>
      <div>
        second inlet
      </div>
    </div>
  </div>
  <div
    id="outlet2"
  >
    <div>
      <div>
        third inlet
      </div>
    </div>
  </div>
</div>
`;

exports[`conduit snapshots Inlet and Outlet should merge the associated Inlet's children to the corresponding outlet 1`] = `
<div>
  <div>
    <div>
      first inlet
    </div>
    <div>
      second inlet
    </div>
    <div>
      third inlet
    </div>
  </div>
</div>
`;

exports[`conduit snapshots Inlet and Outlet should output Inlet's children to the corresponding outlet 1`] = `
<div>
  <div>
    this should be rendered in the outlet
  </div>
</div>
`;

exports[`conduit snapshots Inlet and Outlet two outlets with the same label will duplicate the children rendering 1`] = `
<div>
  <div>
    <div>
      inlet
    </div>
  </div>
  <div>
    <div>
      inlet
    </div>
  </div>
</div>
`;

exports[`conduit snapshots Outlet Outlet should not render its children 1`] = `<div />`;
